### ISSUE TITLE
Issues since 0.11.9 with devise controllers

### ISSUE THREAD

**Issue by fcheung** [Shopify/tapioca#1692]
I'm having the same issue as #1658, namely that there are no longer any rbis generated for controllers.

In particular I have a UserSessionsController that inherits from Devise::UserSessionsController that now produces this error

```
app/controllers/user/sessions_controller.rb:3: Unable to resolve constant SessionsController https://srb.help/5002
     3 |class User::SessionsController < Devise::SessionsController
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  Did you mean User::SessionsController? Use -a to autocorrect
    app/controllers/user/sessions_controller.rb:3: Replace with User::SessionsController
     3 |class User::SessionsController < Devise::SessionsController
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
    app/controllers/user/sessions_controller.rb:3: User::SessionsController defined here
     3 |class User::SessionsController < Devise::SessionsController
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Did you mean Devise::Controllers? Use -a to autocorrect
    app/controllers/user/sessions_controller.rb:3: Replace with Devise::Controllers
     3 |class User::SessionsController < Devise::SessionsController
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
    sorbet/rbi/gems/devise@4.9.3.rbi:1306: Devise::Controllers defined here
    1306 |module Devise::Controllers; end
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

previously an rbi for Devise::SessionsController was generated, which allowed this to work. Devise::SessionsController is normally loaded on demand by rails - adding `Devise::SessionsController` to tapioca/require.rb didn't work ( uninitialized constant Devise::SessionsController)

I've put a small example app with this issue here: https://github.com/fcheung/tapioca_repro

On main, with tapioca 0.11.8  `srb` runs ok, on the tapioca-0.11.9 branch, after running `tapioca dsl`, running `srb` fails with the above.

To summarize - is writing shims what I should do here or is this a regression?

**Comment by KaanOzkan** [comment]
Thanks for the detailed repro. Looks like devise is popular enough where we might want to support it somehow. I think the main issue is `SessionsController` isn't loaded during gem RBI generation.

What do you think of adding the removed definitions from your branch into a file for devise in https://github.com/shopify/rbi-central? I think shims are valid in this case and we can reuse them through rbi-central.

**Comment by andyw8** [comment]
@fcheung you can also try adding this to `require.rb` before generating the RBI:

`$LOAD_PATH.concat(Devise::Engine.config.eager_load_paths)`

The shim may be still be need for some things though.

**Comment by paracycle** [comment]
The root of this problem is this line in Devise: https://github.com/heartcombo/devise/blob/e2242a95f3bb2e68ec0e9a064238ff7af6429545/app/controllers/devise_controller.rb#L4

The default value of `Devise.parent_controller` is `"ApplicationController"` and `.constantize` tries to get the constant with that name. However, Tapioca does NOT load your Rails application as it is doing gem RBI generation, since the assumption is that a gem's RBI should not depend on the application behaviour. After all, the dependency should be one-directional, from the app to the gem.

However, that expectation is broken when we encounter stuff like this where the gem tries to load things that should be a part of the application, so Tapioca fails to generate an RBI for any of Devise controllers.

We could think about starting to load the application (non-eagerly) to allow the initializers to run and for things like `ApplicationController` to have been loaded, but we need to experiment with it to see what that looks like.

In the meantime, the static RBI annotation addition is a good workaround.

**Comment by KaanOzkan** [comment]
@evsasse  I assume Frederick added the constants he was using to rbi-central. Feel free to add any other definitions to https://github.com/Shopify/rbi-central/blob/main/rbi/annotations/devise.rbi if they are referenced in your application.

**Comment by swrobel** [comment]
Why is it necessary to add annotations for all of the various controllers that inherit from `DeviseController`? If that one is annotated to get around this issue, then shouldn't we be golden?

**Comment by KaanOzkan** [comment]
> Why is it necessary to add annotations for all of the various controllers that inherit from DeviseController? If that one is annotated to get around this issue, then shouldn't we be golden?

Yes, no need to add if you don't get type checking errors.
